transports/websys/src/websocket.rs

```diff
@@ -0,0 +1,321 @@
+
+        Ok(async move {
+            let socket = match WebSocket::new(&ws_url) {
+                Ok(ws) => ws,
+                Err(err) => return Err(Error::JsError(format!("{err:?}"))),
+            };
+
+            Ok(Connection::new(socket))
+        }
+        .boxed())
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        _addr: Multiaddr,
+    ) -> Result<Self::Dial, TransportError<Self::Error>> {
+        Err(TransportError::Other(Error::NotSupported))
+    }
+
+    fn poll(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> std::task::Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        Poll::Pending
+    }
+
+    fn address_translation(&self, _listen: &Multiaddr, _observed: &Multiaddr) -> Option<Multiaddr> {
+        None
+    }
+}
+
+// Try to convert Multiaddr to a Websocket url.
+fn websocket_url(addr: Multiaddr) -> Option<String> {
+    let mut protocols = addr.iter();
+    let host_port = match (protocols.next(), protocols.next()) {
+        (Some(Protocol::Ip4(ip)), Some(Protocol::Tcp(port))) => {
+            format!("{ip}:{port}")
+        }
+        (Some(Protocol::Ip6(ip)), Some(Protocol::Tcp(port))) => {
+            format!("[{ip}]:{port}")
+        }
+        (Some(Protocol::Dns(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns4(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dns6(h)), Some(Protocol::Tcp(port)))
+        | (Some(Protocol::Dnsaddr(h)), Some(Protocol::Tcp(port))) => {
+            format!("{}:{}", &h, port)
+        }
+        _ => return None,
+    };
+
+    let (scheme, wspath) = match protocols.next() {
+        Some(Protocol::Ws(path)) => ("ws", path.into_owned()),
+        Some(Protocol::Wss(path)) => ("wss", path.into_owned()),
+        _ => return None,
+    };
+
+    // TODO: handle PeerId
+    Some(format!("{scheme}://{host_port}{wspath}"))
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("js function error {0}")]
+    JsError(String),
+    #[error("operation not supported")]
+    NotSupported,
+}
+
+/// A Websocket connection created by the [`Transport`].
+pub struct Connection {
+    /// We need to use Mutex as libp2p requires this to be Send.
+    shared: Arc<Mutex<Shared>>,
+}
+
+struct Shared {
+    opened: bool,
+    closed: bool,
+    error: bool,
+    data: VecDeque<u8>,
+    waker: Option<Waker>,
+    socket: SendWrapper<WebSocket>,
+    closures: Option<SendWrapper<Closures>>,
+}
+
+type Closures = (
+    Closure<dyn FnMut()>,
+    Closure<dyn FnMut(MessageEvent)>,
+    Closure<dyn FnMut(web_sys::Event)>,
+    Closure<dyn FnMut(web_sys::CloseEvent)>,
+);
+
+impl Connection {
+    fn new(socket: WebSocket) -> Self {
+        socket.set_binary_type(web_sys::BinaryType::Arraybuffer);
+
+        let shared = Arc::new(Mutex::new(Shared {
+            opened: false,
+            closed: false,
+            error: false,
+            data: VecDeque::with_capacity(1 << 16),
+            waker: None,
```

Comment:
I am not sure it is good to only have a single waker here. I'd err on the side of caution and split these per `poll_` function.